### PR TITLE
Store heading tree and display sidebar for content heavy pages (fixes #272)

### LINKED ISSUE #272: Implement sidebar for content heavy pages
A sidebar similar to the CSFG would be useful for the following pages:

- Unit Plan
- Lesson
- Programming Exercise

### PR DESCRIPTION
**Pull request by JackMorganNZ**
## Proposed changes

This pull request:

- Store the heading tree for Lessons in JSON field.
- Display sidebar on lesson page.
- Use sticky positioning of sidebar.
- Use scrollspy to highlight position on sidebar.

The JSON field is specific to Postgres databases, however Postgres is feature rich, well supported, free, and open source, so we should be safe with using a Postgres specific field. It can be removed in a emergency to allow generic database use.

I am still yet to add to this pull request:

- Deployment documentation stating Postgres database is required.

This pull request is **NOT** for:

- Finalizing the design (size, colour, etc) of the sidebar.
- Storing the heading tree for all required models.
- Adding the sidebar to all required pages.
- Improving the template for creating the sidebar.
- Mobile interface for sidebar.

This pull request is more a proof of concept, which will be developed further in other pull requests if approved.

These features are added in other issues for the "Website Interface Update - Round 3" milestone.

### CODE REVIEW

File: csunplugged/utils/convert_heading_tree_to_dict.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    Returns:
+        Dictionary of heading tree, or None if None provided as tree.
+    """
+    if heading_tree_tuples is None:
+        return None
+    else:
+        root_nodes = []
+        for heading_node in heading_tree_tuples:
+            root_nodes.append(convert_heading_node(heading_node))
+        return root_nodes
```
Comment (ravenmaster001) [inline csunplugged/utils/convert_heading_tree_to_dict.py]: `return [convert_heading_node(heading_node) for heading_node in heading_tree_tuples]`

File: csunplugged/utils/convert_heading_tree_to_dict.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+
+def convert_heading_node(heading_node):
```
Comment (ravenmaster001) [inline csunplugged/utils/convert_heading_tree_to_dict.py]: Give this a better name please.

File: csunplugged/static/js/fixedsticky.js [hunk 3]
```diff
@@ -0,0 +1,193 @@
+;(function( win, $ ) {
```
Comment (ravenmaster001) [inline csunplugged/static/js/fixedsticky.js]: I know this is probably third-party but I feel we should have a comment at the top of scripts or some documentation explaining what functionality each javascript file provides. I always feel that javascript libraries/scripts tend to be poorly named and are not immediately obvious what they are used for let alone how to use them.

File: csunplugged/static/scss/sticky-state.scss [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  position: -webkit-sticky;
+  position: sticky;
+}
+.sticky.sticky-fixed.is-sticky {
+  margin-top: 0;
+  margin-bottom: 0;
+  position: fixed;
+  -webkit-backface-visibility: hidden;
+  -moz-backface-visibility: hidden;
+  backface-visibility: hidden;
```
Comment (ravenmaster001) [inline csunplugged/static/scss/sticky-state.scss]: weird codehilighting

Comment (JackMorganNZ) [inline csunplugged/static/scss/sticky-state.scss]: That is weird, might be because it's not supported without prefix currently?

Comment (JackMorganNZ) [inline csunplugged/static/scss/sticky-state.scss]: Yep, looks like it's currently only supported with prefix (https://developer.mozilla.org/en-US/docs/Web/CSS/backface-visibility#Browser_compatibility) so this is a future proofed CSS rule. Should be fine to leave as is.

**Pull request merged**